find/client/http/dhash_client.go

```diff
@@ -70,101 +71,115 @@ func NewDHashClient(dhstoreURL, stiURL string, options ...Option) (*DHashClient,
+		case err := <-errChan:
+			return nil, err
+		case pr, ok := <-resChan:
+			if !ok {
+				return &model.FindResponse{
+					MultihashResults: []model.MultihashResult{mhr},
+				}, nil
+			}
+			mhr.ProviderResults = append(mhr.ProviderResults, pr)
+		}
+	}
+}
+
+// FindAsync implements double hashed lookup workflow. It submits results as they get decrypted and assembled into resChan. If an error occurs it is sent to errChan.
+// Once the workflow is finished both channels are closed.
+func (c *DHashClient) FindAsync(ctx context.Context, mh multihash.Multihash, resChan chan model.ProviderResult, errChan chan error) {
+	defer func() {
+		close(resChan)
+		close(errChan)
+	}()
+
 	smh, err := dhash.SecondMultihash(mh)
 	if err != nil {
-		return nil, err
+		errChan <- err
 	}
 	u := c.dhFindURL.JoinPath(smh.B58String())
 	req, err := http.NewRequestWithContext(ctx, http.MethodGet, u.String(), nil)
 	if err != nil {
-		return nil, err
+		errChan <- err
 	}
 	req.Header.Add("Accept", "application/json")
 
 	resp, err := c.c.Do(req)
 	if err != nil {
-		return nil, err
+		errChan <- err
 	}
 
 	body, err := io.ReadAll(resp.Body)
 	defer resp.Body.Close()
 
 	if err != nil {
-		return nil, err
+		errChan <- err
 	}
 
 	if resp.StatusCode != http.StatusOK {
-		return nil, apierror.FromResponse(resp.StatusCode, body)
+		errChan <- apierror.FromResponse(resp.StatusCode, body)
 	}
 
-	findResponse := &model.FindResponse{}
-	err = json.Unmarshal(body, findResponse)
+	encResponse := &model.FindResponse{}
+	err = json.Unmarshal(body, encResponse)
 	if err != nil {
-		return nil, err
+		errChan <- err
 	}
 
-	err = c.decryptFindResponse(ctx, findResponse, map[string]multihash.Multihash{smh.B58String(): mh})
-	if err != nil {
-		return nil, err
-	}
-
-	return findResponse, nil
-}
-
-// decryptFindResponse decrypts EncMultihashResults and appends the decrypted
-// values to MultihashResults. It also fetches provider info and metadata for
-// each value key.
-func (c *DHashClient) decryptFindResponse(ctx context.Context, resp *model.FindResponse, unhasher map[string]multihash.Multihash) error {
-	// decrypt each value key using the original multihash
-	// then for each decrypted value key fetch provider's addr info.
-	for _, encRes := range resp.EncryptedMultihashResults {
-		mh, found := unhasher[encRes.Multihash.B58String()]
-		if !found {
-			continue
-		}
-
-		mhr := model.MultihashResult{
-			Multihash: mh,
-		}
-		for _, evk := range encRes.EncryptedValueKeys {
-			vk, err := dhash.DecryptValueKey(evk, mh)
-			// skip errors as we don't want to fail the whole query, warn instead. Same applies to the rest of the loop.
-			if err != nil {
-				log.Warnw("Error decrypting value key", "multihash", mh.B58String(), "evk", b58.Encode(evk), "err", err)
-				continue
+	for _, emhrs := range encResponse.EncryptedMultihashResults {
+		for _, evk := range emhrs.EncryptedValueKeys {
+			select {
+			case <-ctx.Done():
+				return
+			default:
```

Comment:
Instead of using `select` to check for the context channel being closed and `default`, suggest simply checking the context for error.
```suggestion
			if ctx.Err() != nil {
				errChan <- ctx.Err()
				return
			}
```
It is [synchronized](https://cs.opensource.google/go/go/+/refs/tags/go1.20.3:src/context/context.go;l=429-431) if that was a concern.